### ISSUE TITLE
Access URLs and services list stuck in loading 

### ISSUE THREAD

**Issue by migmartri** [kubeapps/kubeapps#882]
If you deploy an application and then delete for example a service, the UI will get stuck in loading state.

This is likely due to us waiting for websockets to return state but this never arrives. 

![kubeapps dashboard](https://user-images.githubusercontent.com/24523/50106262-fbd1f600-01f4-11e9-9762-21556589fe44.png)

**Comment by migmartri** [comment]
Note that `helm status` reports the service from being missing 
```
==> MISSING
KIND              NAME
{ v1 services}    turbulent-grizzly-ghost
 ```

**Comment by migmartri** [comment]
Same problem appears if you see a release that has been deleted but not purged.

**Comment by prydonius** [comment]
Hmm, good find. I'm not sure if there's a way from the WebSockets that we can tell if the resource was actually found or not (since we're watching the list endpoint and we only get an event if it exists). I guess there are two options I can think of:

- Have a timeout for the loading indicator, if after 30s we don't hear back, just show nothing
- Make a regular HTTP request for the resource before/in parallel to the WebSocket to check for a 404 response

What should the UX be if a resource isn't found? We can warn that the release and actual state is inconsistent and prompt the user to upgrade the chart, for example.

**Comment by andresmgot** [comment]
> Make a regular HTTP request for the resource before/in parallel to the WebSocket to check for a 404 response

My +1 to that. That's how the SecretsTable behaves. The problem is when more than one element is missing, in that case, right now, it will only show the latest error received.

I wouldn't do the timeout thing because for a deleted app it would be weird to wait 30 s to get the notification (which should be immediate).

Ideally I think that we should parse the elements from the release manifest, if a 404 is received when trying to get the object we could add a warning icon that the item is missing.

**Comment by migmartri** [comment]
There might be a way to sanitize the manifest before we subscribe to the different resources. With Helm status we get
 
```==> MISSING
KIND              NAME
{ v1 services}    turbulent-grizzly-ghost`
```

so there might be a way for us to remove the resources in the interception of those two sets.

**Comment by migmartri** [comment]
w.r.t making a request from the frontend instead of waiting for the ws to return, I think it can make sense since at least for me, the websocket handshake and response seems quite slow compared to the http requests for the secrets.

**Comment by prydonius** [comment]
> There might be a way to sanitize the manifest before we subscribe to the different resources. With Helm status we get

Parsing Helm status is going to be really hacking, so I think actually making the request will be a lot more robust. This is effectively what Helm status is doing under the hood anyway to find out what is missing.

**Comment by prydonius** [comment]
This might be a good opportunity to refactor the AppView into containers, similar to SecretsTable, where we can deal with the request flow for the resources. In these containers, we can first make a GET request to fetch the resources, once successfully loaded we can open WebSockets for them. Another option is to perform the GET request and open the WebSocket at the same time, and display whatever gets back sooner or an error if the GET request fails. What do you think @migmartri?

**Comment by migmartri** [comment]
> This might be a good opportunity to refactor the AppView into containers, similar to SecretsTable, where we can deal with the request flow for the resources. In these containers, we can first make a GET request to fetch the resources, once successfully loaded we can open WebSockets for them. Another option is to perform the GET request and open the WebSocket at the same time, and display whatever gets back sooner or an error if the GET request fails.
> 
> Currently we open a WebSocket per-resource, filtering using `metadata.name`. It might be more efficient to open a single WebSocket per-kind (e.g. Deployments) and cache each resource (potentially even things we don't need - though we can also filter those out, somehow) in `IKubeState`. The `IKubeState` can track which sockets are open and handle closing them when the relevant component dismounts.
> 
> What do you think @migmartri?

I agree with this approach :)

**Comment by migmartri** [comment]
For me, fetching resources and the websockets are two different things so they can be implemented individually.

1 - Refactor to fetch the resources by following a container-component approach and relying on the redux store (similarly to what we do with the secrets resources).
2 - Implement a generic watcher component that opens and closes websockets to specific resource types. Note that this watcher must not rely on local storage but instead dispatch actions with the changes in the subscribed resources. Ideally it should only open a websocket per kind not per resource.

By implementing only 1 we will have fixed the issue described in this thread and will not add a functionality regression. So I'd start with that and then continue with the second step.

**Comment by migmartri** [comment]
Actually today during testing I noticed that this issue occurred to me even in the case of nothing has being deleted. 

As you can see below, all websockets communication were handshaked (101) but no response was received (I waited a couple of minutes and no luck).

![selection_797](https://user-images.githubusercontent.com/24523/50525360-d7e1a080-0a8f-11e9-9dcc-148120318b9d.png)

I'd like to bump the priority of this task to implement the first step described here https://github.com/kubeapps/kubeapps/issues/882#issuecomment-450399977 as soon as possible since this page at least to me feels buggy and unresponsive. WDYT?

**Comment by prydonius** [comment]
@migmartri agreed, I think we should implement that first step in this iteration.

**Comment by prydonius** [comment]
### State changes

We need to extend IKubeState to store open sockets for resources, following the pattern from (https://github.com/kubeapps/kubeapps/issues/912) we will create a new type `IKubeRefs` which sets up the GVK -> namespace -> resource hierarchy and use that to store WebSockets in IKubeState:

```
+export interface IKubeRefs<T> {
+  // groupVersionKind (e.g. v1/secrets, apps/v1/deployments, kubeapps.com/v1alpha1/apprepositories)
+  [groupVersionKind: string]: {
+    [namespace: string]: {
+      [name: string]: T;
+    };
+  };
+}
+
 export interface IKubeState {
   items: { [s: string]: IKubeItem<IResource> };
+  sockets: IKubeRefs<WebSocket>;
 }
```

### Actions

#### getAndWatchResource

We will add an action similar to https://github.com/kubeapps/kubeapps/blob/c8bb6ea82c7304783835a79476f08c51a324c5c4/dashboard/src/actions/kube.tsx#L22 that will asynchronously dispatch `getResource` and at the same time open a WebSocket connection (if one does not exist).The WebSocket will be configured with a handler will dispatch a `receiveResource` action each time a message is received with the resource, thus updating the Redux store. The returned socket will be dispatched in a `receiveSocket` action that passes the groupVersionKind, namespace, name and socket to the reducer. 

#### closeWatchResource

We will add an action that takes a groupVersionKind, namespace and name to find an open WebSocket connection in the state and close it out. It will dispatch a `removeSocket` action.

#### receiveSocket

Dispatched by `getAndWatchResource` when a WebSocket is opened.

#### removeSocket

Dispatched by `closeWatchResource` when a WebSocket is closed.


### Reducer

#### receiveSocket

Adds the socket to the IKubeState under the appropriate reference.

#### removeSocket

Removes the socket in the IKubeState under the reference.

**Comment by prydonius** [comment]
### Container/Component changes

#### Wrap tables in Containers

The `DeploymentsTable` and `ServicesTable` will be wrapped into Containers, similarly to `SecretsTable`. These Containers will be responsible for mapping the `getAndWatchResource` and `closeWatchResource` actions to the Components so that they can be called during the lifecycle.

The Containers will also pass down the resources related to the currently selected App down to the Components. To do so, the Containers will take a prop containing a list of resource IDs (namespace/name) that it will then use to filter the IKubeState for the resources. This leaves the parsing of the manifest as-is in the AppView (https://github.com/kubeapps/kubeapps/blob/c8bb6ea82c7304783835a79476f08c51a324c5c4/dashboard/src/components/AppView/AppView.tsx#L104), however in the future this manifest could be parsed in the reducer and information extracted can be stored in the Redux state.

#### DeploymentStatus and DeploymentsTable

The `DeploymentStatus` Component also needs to read the latest information about the Deployments. There are two options here:

1. Assume that the WebSocket connection is handled by `DeploymentsTable` and just wrap in a Container that hands down the Deployments of the App
2. Don't assume 1, instead wrap in a Container that provides `getAndWatchResource` and `closeWatchResource` and assume that the actions handle the case where a WebSocket is already open for a resource (this could easily lead to race conditions, so might not be possible)

It might be simplest for now to just go with option 1, where we only handle the WebSocket connection in `DeploymentsTable`, and revisit this when we want a higher-order `ResourceWatcher` Container.

**Comment by migmartri** [comment]
> getAndWatchResource

Something that we hinted was the possibility of watching multiple resources in a single WS connection, how could we handle this? I am wondering if this action should not open the connection but just store it in Redux. Then have another container that will take care of grouping those references and open connections. 

> Wrap tables in Containers

Are these going to retrieve all elements at once? Otherwise the container is around an item not the whole table isn't it?

> DeploymentStatus and DeploymentsTable

For simplicity I'd go for option 1.

What would be the action plan here? To do both the container/component changes and the new websockets lifecycle? Could we start for now with just the container/component part? 

Thanks for the write up!

**Comment by prydonius** [comment]
> Something that we hinted was the possibility of watching multiple resources in a single WS connection, how could we handle this? I am wondering if this action should not open the connection but just store it in Redux. Then have another container that will take care of grouping those references and open connections.

I wanted to keep things similar to what we have right now, so currently this will open a watch with the metadata.name filter for the specific resource. In the follow up issue, we can change this to simply open a watch for all resources in the namespace and possibly a list of names to cache client side. This would require changing the redux store a bit again (IKubeRef will no longer be needed for sockets, we will just need one for each groupVersionKind and namespace, e.g. `sockets: { [groupVersionKindNamespace: string]: WebSocket }`), and then update `getAndWatchResource` to open the socket if not exists and (if needed) add the resource name to watch to a list. This change definitely needs some more thought though, as I think there are side effects to changing individual watches to a single watch (e.g. how do you know when to close the watch). For now, it's simpler to keep things the same and just move the socket handling to the redux store.

> Are these going to retrieve all elements at once? Otherwise the container is around an item not the whole table isn't it?

Good point, I was going by the SecretsTable which loops through the Secrets and fetches them, but I agree it makes much more sense to do this in the Item components. I'll update the proposal to do so, and we should probably create an issue to refactor the SecretsTable to do it this way too.

> What would be the action plan here? To do both the container/component changes and the new websockets lifecycle? Could we start for now with just the container/component part?

I will try to work on the container/component changes in a separate PR, though without moving the WebSockets to the Redux store, we can't perform the HTTP GET to fix the issue without handling the case where data can come from different places (WebSocket from AppsView or Redux store via Container). I'll try to think of a good way of separating the pieces so this is easier to review.

**Comment by prydonius** [comment]
After discussing with @migmartri, we can split this work by refactoring into the Containers/Components and changing the WebSocket handlers to push to the Redux store, but keep them in the AppsView.

**Comment by andresmgot** [comment]
I am okay with the approach, just one small comment for:

> Something that we hinted was the possibility of watching multiple resources in a single WS connection,

The user token may don't have permissions to list resources (like Secrets) but to get specific ones so I would rather do the request in a component basis.

**Comment by shangjunimo** [comment]
![image](https://user-images.githubusercontent.com/16662815/51593309-d7bcaf80-1f2c-11e9-8ac8-2f722a9cf379.png)
The problem I encountered was 401, and I didn't know how to solve it.
